Allow custom registration of shape functions

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #73329
* __->__ #73270
* #69871

Together with open registration of NNC lowerings this should make possible to add support for custom operators, including internal fb-ops

Differential Revision: [D34451275](https://our.internmc.facebook.com/intern/diff/D34451275)

Commit:
Allow custom registration of shape functions

[ghstack-poisoned]

Commit:
Update on "Allow custom registration of shape functions"




[ghstack-poisoned]

davidberard98 commented on torch/csrc/jit/runtime/symbolic_shape_registry.cpp:
```diff
@@ -292,5 +292,14 @@ c10::optional<std::shared_ptr<Graph>> shapeComputeGraphForSchema(
   return c10::nullopt;
 }
 
+void RegisterShapeComputeGraphForSchema(const FunctionSchema& schema, std::shared_ptr<Graph> g) {
```
nit: maybe take a `FunctionSchema*` since that's what's needed in the map? e.g. it might prevent someone from making an accidental copy if it's clear that the function takes a pointer as an argument.

also, just curious - do you know how the FunctionSchema lifetimes are handled? Seems like a lot of code for FunctionSchemas is written like this so clearly it works... is there a global cache somewhere that returns the same FunctionSchema object every time it sees the same schema?

eellison replied:
There is this comment about a static lifetime of operators, and corresponding static lifetime of function schema.. didnt look into it further yet but that file should b enough to figure things out

https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/runtime/operator.h#L36

davidberard98 reviewed (approved):
looks good!

Commit:
Update on "Allow custom registration of shape functions"


Together with open registration of NNC lowerings this should make possible to add support for custom operators, including internal fb-ops

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34451275).

Commit:
Update on "Allow custom registration of shape functions"


Together with open registration of NNC lowerings this should make possible to add support for custom operators, including internal fb-ops

Differential Revision: [D34451275](https://our.internmc.facebook.com/intern/diff/D34451275)

[ghstack-poisoned]

Commit:
Update on "Allow custom registration of shape functions"


Together with open registration of NNC lowerings this should make possible to add support for custom operators, including internal fb-ops

Differential Revision: [D34451275](https://our.internmc.facebook.com/intern/diff/D34451275)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34451275).

Commit:
Update on "Allow custom registration of shape functions"


Together with open registration of NNC lowerings this should make possible to add support for custom operators, including internal fb-ops

Differential Revision: [D34451275](https://our.internmc.facebook.com/intern/diff/D34451275)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34451275).

Commit:
Update on "Allow custom registration of shape functions"


Together with open registration of NNC lowerings this should make possible to add support for custom operators, including internal fb-ops

Differential Revision: [D34451275](https://our.internmc.facebook.com/intern/diff/D34451275)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34451275).
